// vuetify/es5/components/VTabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "VTabs", {
  enumerable: true,
  get: function get() {
    return _VTabs.default;
  }
});
Object.defineProperty(exports, "VTab", {
  enumerable: true,
  get: function get() {
    return _VTab.default;
  }
});
Object.defineProperty(exports, "VTabsItems", {
  enumerable: true,
  get: function get() {
    return _VTabsItems.default;
  }
});
Object.defineProperty(exports, "VTabItem", {
  enumerable: true,
  get: function get() {
    return _VTabItem.default;
  }
});
Object.defineProperty(exports, "VTabsSlider", {
  enumerable: true,
  get: function get() {
    return _VTabsSlider.default;
  }
});
exports.default = void 0;

var _VTabs = _interopRequireDefault(require("./VTabs"));

var _VTab = _interopRequireDefault(require("./VTab"));

var _VTabsItems = _interopRequireDefault(require("./VTabsItems"));

var _VTabItem = _interopRequireDefault(require("./VTabItem"));

var _VTabsSlider = _interopRequireDefault(require("./VTabsSlider"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  $_vuetify_subcomponents: {
    VTabs: _VTabs.default,
    VTab: _VTab.default,
    VTabsItems: _VTabsItems.default,
    VTabItem: _VTabItem.default,
    VTabsSlider: _VTabsSlider.default
  }
};
exports.default = _default;
