// vuetify/es5/components/VGrid/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "VContainer", {
  enumerable: true,
  get: function get() {
    return _VContainer.default;
  }
});
Object.defineProperty(exports, "VCol", {
  enumerable: true,
  get: function get() {
    return _VCol.default;
  }
});
Object.defineProperty(exports, "VRow", {
  enumerable: true,
  get: function get() {
    return _VRow.default;
  }
});
Object.defineProperty(exports, "VSpacer", {
  enumerable: true,
  get: function get() {
    return _VSpacer.default;
  }
});
Object.defineProperty(exports, "VLayout", {
  enumerable: true,
  get: function get() {
    return _VLayout.default;
  }
});
Object.defineProperty(exports, "VFlex", {
  enumerable: true,
  get: function get() {
    return _VFlex.default;
  }
});
exports.default = void 0;

var _VContainer = _interopRequireDefault(require("./VContainer"));

var _VCol = _interopRequireDefault(require("./VCol"));

var _VRow = _interopRequireDefault(require("./VRow"));

var _VSpacer = _interopRequireDefault(require("./VSpacer"));

var _VLayout = _interopRequireDefault(require("./VLayout"));

var _VFlex = _interopRequireDefault(require("./VFlex"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  $_vuetify_subcomponents: {
    VContainer: _VContainer.default,
    VCol: _VCol.default,
    VRow: _VRow.default,
    VSpacer: _VSpacer.default,
    VLayout: _VLayout.default,
    VFlex: _VFlex.default
  }
};
exports.default = _default;
