// vuetify/es5/components/VDatePicker/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "VDatePicker", {
  enumerable: true,
  get: function get() {
    return _VDatePicker.default;
  }
});
Object.defineProperty(exports, "VDatePickerTitle", {
  enumerable: true,
  get: function get() {
    return _VDatePickerTitle.default;
  }
});
Object.defineProperty(exports, "VDatePickerHeader", {
  enumerable: true,
  get: function get() {
    return _VDatePickerHeader.default;
  }
});
Object.defineProperty(exports, "VDatePickerDateTable", {
  enumerable: true,
  get: function get() {
    return _VDatePickerDateTable.default;
  }
});
Object.defineProperty(exports, "VDatePickerMonthTable", {
  enumerable: true,
  get: function get() {
    return _VDatePickerMonthTable.default;
  }
});
Object.defineProperty(exports, "VDatePickerYears", {
  enumerable: true,
  get: function get() {
    return _VDatePickerYears.default;
  }
});
exports.default = void 0;

var _VDatePicker = _interopRequireDefault(require("./VDatePicker"));

var _VDatePickerTitle = _interopRequireDefault(require("./VDatePickerTitle"));

var _VDatePickerHeader = _interopRequireDefault(require("./VDatePickerHeader"));

var _VDatePickerDateTable = _interopRequireDefault(require("./VDatePickerDateTable"));

var _VDatePickerMonthTable = _interopRequireDefault(require("./VDatePickerMonthTable"));

var _VDatePickerYears = _interopRequireDefault(require("./VDatePickerYears"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  $_vuetify_subcomponents: {
    VDatePicker: _VDatePicker.default,
    VDatePickerTitle: _VDatePickerTitle.default,
    VDatePickerHeader: _VDatePickerHeader.default,
    VDatePickerDateTable: _VDatePickerDateTable.default,
    VDatePickerMonthTable: _VDatePickerMonthTable.default,
    VDatePickerYears: _VDatePickerYears.default
  }
};
exports.default = _default;
